File: C++/sortings/mergesort.cpp
#include <vector>
#include <iostream>

std::vector<int> merge(const std::vector<int>& v1, const std::vector<int>& v2){
    int szr = 0;
    int sz1 = (int) v1.size();
    int sz2 = (int) v2.size();
    int p1 = 0;
    int p2 = 0;
    std::vector<int> result(sz1 + sz2);
    while (szr <= sz1 + sz2 - 1){
        if (p1 == sz1 or (p2 < sz2 and v1.at(p1) > v2.at(p2))){
            result.at(szr) = v2.at(p2);
            p2++;
        }
        else{
            result.at(szr) = v1.at(p1);
            p1++;
        }
        szr++;
    }
    return result;
}

std::vector<int> mergesort(std::vector<int> arr){
    if (arr.size() == 1){
        return arr;
    }
    else{
        int sz = (int) arr.size();
        std::vector<int> v1(sz / 2);
        std::vector<int> v2(sz - (sz / 2));
        for (int i = 0; i < sz; i++){
            if (i < sz / 2){
                v1.at(i) = arr.at(i);
            }
            else{
                v2.at(i - (sz / 2)) = arr.at(i);
            }
        }
        return merge(mergesort(v1), mergesort(v2));
    }
}

int main(){
    int t;
    std::cin >> t;
    std::vector<int> v(t);
    for (int i = 0; i < t; i++){
        std::cin >> v[i];
    }
    std::vector<int> sorted = mergesort(v);
    
    for (int i = 0; i < t; i++){
        printf("%d ", sorted.at(i));
    }

    return 0;

}

File: Python/sortings/mergesort.py
def merge(v1: list[int], v2: list[int]) -> list[int]:
    result = [0] * (len(v1) + len(v2))
    p1 = 0
    p2 = 0
    for i in range(len(v1) + len(v2)):
        if p1 == len(v1) or p2 < len(v2) and v1[p1] > v2[p2]:
            result[i] = v2[p2]
            p2 += 1
        else:
            result[i] = v1[p1]
            p1 += 1
    return result


def merge_sort(v: list[int]) -> list[int]:
    if len(v) == 1:
        return v
    v1 = v[:len(v) // 2]
    v2 = v[len(v) // 2:]
    return merge(merge_sort(v1), merge_sort(v2))


print(*merge_sort(list(map(int, input().split()))))
